File: python/460_LFUCache.py
from unittest import TestCase
# https://leetcode.com/problems/lfu-cache/


class Node:
    def __init__(self, key, value, freq=1):
        self.key = key
        self.value = value
        self.freq = freq
        self.prev = self.next = None

    def append(self, node):
        if self.next is not None:
            self.next.prev = node
        node.next = self.next
        node.prev = self
        self.next = node

    def remove(self):
        self.next.prev = self.prev
        self.prev.next = self.next


class LFUCache:
    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity
        self.count = 0
        self.min = -1
        self.freq_map = dict()
        self.node_map = dict()

    def get_freq(self, freq):
        if freq not in self.freq_map:
            head = Node(-1, -1)
            tail = Node(-1, -1)
            head.append(tail)
            self.freq_map[freq] = (head, tail)
        return self.freq_map[freq]

    def remove_node(self, node):
        node.remove()
        del self.node_map[node.key]
        head, tail = self.get_freq(node.freq)
        if head.next == tail and self.min == node.freq:
            self.min += 1
        self.count -= 1

    def insert_node(self, node):
        head, tail = self.get_freq(node.freq)
        head.append(node)
        self.node_map[node.key] = node
        self.count += 1

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key not in self.node_map:
            return -1
        node = self.node_map[key]
        self.remove_node(node)

        node.freq += 1
        self.insert_node(node)
        return node.value

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: void
        """
        if self.capacity == 0:
            return
        if key in self.node_map:
            self.node_map[key].value = value
            return self.get(key)

        if self.count == self.capacity:
            head, tail = self.get_freq(self.min)
            self.remove_node(tail.prev)

        node = Node(key, value, 1)
        self.insert_node(node)
        self.min = 1


class Tests(TestCase):
    def test1(self):
        cache = LFUCache(0)
        cache.put(1, 1)
        self.assertEqual(-1, cache.get(1))

    def test2(self):
        cache = LFUCache(3)
        cache.put(2, 2)
        cache.put(1, 1)
        self.assertEqual(2, cache.get(2))
        self.assertEqual(1, cache.get(1))
        self.assertEqual(2, cache.get(2))
        cache.put(3, 3)
        cache.put(4, 4)
        self.assertEqual(-1, cache.get(3))
        self.assertEqual(2, cache.get(2))
        self.assertEqual(1, cache.get(1))
        self.assertEqual(4, cache.get(4))
        cache.put(2, 9)
        self.assertEqual(9, cache.get(2))

File: kotlin/src/main/kotlin/460.LFUCache.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/lfu-cache/


class LFUCache(val capacity: Int) {
  class Node<T>(val value: T){
    var prev: Node<T>?=null
    var next: Node<T>?=null

    fun append(node: Node<T>){
      node.next = next
      next?.prev = node
      next = node
      node.prev = this
    }

    fun remove(){
      prev?.next = next
      next?.prev = prev
    }
  }

  data class Data(val key: Int, var value: Int, var freq: Int=1)
  data class FreqInfo(val freq: Int, val dataHead: Node<Data>, val dataTail: Node<Data>){
    init { dataHead.append(dataTail) }
  }

  private fun buildFreqNode(freq: Int) = Node(FreqInfo(freq, Node(Data(-1, -1)), Node(Data(-1, -1))))

  var count = 0
  val map = HashMap<Int, Node<Data>>()
  val freqMap = HashMap<Int, Node<FreqInfo>>()
  val freqHead = buildFreqNode(-1)

  fun get(key: Int): Int {
    if(!map.contains(key)) return -1
    val dataNode = map[key]!!
    val prevFreqNode = remove(dataNode)
    dataNode.value.freq ++
    insert(dataNode, prevFreqNode)
    return dataNode.value.value
  }

  fun put(key: Int, value: Int) {
    if(capacity == 0) return
    if(map.containsKey(key)){
      val node = map[key]!!
      node.value.value = value
      get(key)
    } else {
      if (count == capacity) pop()
      val node = Node(Data(key, value))
      insert(node)
    }
  }

  private fun emptyFreq(freqInfo: FreqInfo) = freqInfo.dataHead.next == freqInfo.dataTail
  private fun remove(dataNode: Node<Data>): Node<FreqInfo>{  // return prev freqNode
    dataNode.remove()
    map.remove(dataNode.value.key)
    count -= 1

    val freq = dataNode.value.freq
    val freqNode = freqMap[freq]!!
    if(emptyFreq(freqNode.value)) {
      freqNode.remove()
      freqMap.remove(freq)
      return freqNode.prev!!
    }else
      return freqNode
  }
  private fun insert(node: Node<Data>, prevFreqNode: Node<FreqInfo> = freqHead){
    assert(!map.contains(node.value.key))
    map[node.value.key] = node
    count += 1

    val freq = node.value.freq
    if(!freqMap.contains(freq))
      freqMap[freq] = buildFreqNode(freq).also{ prevFreqNode.append(it) }
    val freqNode = freqMap[freq]!!
    freqNode.value.dataHead.append(node)
  }

  private fun pop(){
    val leastFreqNode = freqHead.next!!
    val dataNode = leastFreqNode.value.dataTail.prev!!
    remove(dataNode)
  }

}

class LFUCacheTests{
  @Test
  fun test1(){
    val cache = LFUCache( 2)
    cache.put(1, 1)
    cache.put(2, 2)
    assertEquals(1, cache.get(1))
    cache.put(3, 3)
    assertEquals(-1, cache.get(2))
    assertEquals(3, cache.get(3))
    cache.put(4, 4)
    assertEquals(-1, cache.get(1))
    assertEquals(3, cache.get(3))
    assertEquals(4, cache.get(4))
  }

  @Test
  fun test2(){
    val cache = LFUCache( 3)
    cache.put(2, 2)
    cache.put(1, 1)
    assertEquals(2, cache.get(2))
    assertEquals(1, cache.get(1))
    assertEquals(2, cache.get(2))
    cache.put(3, 3)
    cache.put(4, 4)
    assertEquals(-1, cache.get(3))
    assertEquals(2, cache.get(2))
    assertEquals(1, cache.get(1))
    assertEquals(4, cache.get(4))
  }
}